unit_scaling/transforms/_unit_scale.py

```diff
@@ -0,0 +1,294 @@
+            if node.op == "call_function":
+                if node.target in replacement_map:
+                    target_fn = replacement_map[node.target]
+                    logger.info(
+                        "replacing function: %s with %s", node, target_fn.__name__
+                    )
+                    replace_node_with_function(graph, node, target_fn)
+                elif node.target in U.torch_map:
+                    target_fn = U.torch_map[node.target]
+                    logger.info("unit scaling function: %s", node)
+                    replace_node_with_function(graph, node, target_fn)
+
+        # Add metadata denoting the dependencies of every node in the graph
+        for node in graph.nodes:
+            if node.op == "output":
+                _add_dependency_meta(node)
+
+        # Go through and mark nodes which represent residual-adds
+        residual_layer_number = 1
+        for node in graph.nodes:
+            if _is_add(node):
+                is_residual_add = False
+                if len(node.args) == 2:
+                    l, r = node.args
+                    if isinstance(l, Node) and isinstance(r, Node):
+                        l_deps = l.meta.get("dependencies", list())
+                        r_deps = r.meta.get("dependencies", list())
+                        if l in r_deps or r in l_deps:
+                            node.meta["residual_add"] = {
+                                "residual_arg_idx": 1 if l in r_deps else 0,
+                                "layer_number": residual_layer_number,
+                            }
+                            residual_layer_number += 1
+                            is_residual_add = True
+                # Regular adds are not picked up by the unit scaling sweep above as
+                # the inbuilt + operation is handled differently when traced. It is
+                # instead substituted for its unit scaled equivalent here.
+                if not is_residual_add:
+                    logger.info("unit scaling function: %s", node)
+                    args = (*node.args, None)  # None denotes unconstrained
+                    replace_node_with_function(graph, node, U.add, args=args)
+        num_residuals = residual_layer_number - 1
+
+        # Replace nodes marked as residual-adds with unit scaled equivalent
+        final_residual_reached = num_residuals == 0
+        for node in graph.nodes:
+            residual_add = node.meta.get("residual_add", None)
+            if residual_add is not None:
+                logger.info("unit scaling function: %s (residual-add)", node)
+                _unit_scale_residual(graph, node, **residual_add)
+                if residual_add["layer_number"] == num_residuals:
+                    final_residual_reached = True
+            elif final_residual_reached:
+                node.meta["after_final_residual"] = True
+
+        for node in graph.nodes:
+            if "after_final_residual" in node.meta:
```

Comment:
I wonder if setting "has_residual_successor" for all "dependencies" of "residual_add" nodes might be better? The "after_final_residual" approach allows some variation of behaviour due to possible toposorts of the graph, I think?

Reply:
Yeah I didn't do the cut-edges thing very thoroughly because as I chop-and-change the graph I don't maintain the dependencies metadata properly and implementing that would be a bit of a pain. But it now strikes me that I could just re-run the original dependency analysis code before this step. Might have a go...

Reply:
Implemented your suggestion - wasn't too bad in the end!

Reply:
Ah, cool! (Sorry, didn't mean to give you more work, should've said it could be a note for later.)